vpc/service/gc_v3.go

```diff
@@ -106,9 +106,123 @@ func (vpcService *vpcService) GCV3(ctx context.Context, req *vpcapi.GCRequestV3)
 		tracehelpers.SetStatus(err, span)
 		return nil, err
 	}
+	if len(resp.AssignmentsToRemove) > 0 {
+		logger.G(ctx).WithField("assignments", resp.AssignmentsToRemove).Info("Fetched assignments to remove")
+	}
 	return resp, nil
 }
 
+// Set gc_tombstone time to now() on any assginments that match the following condition:
+// 1) The assignment uses the given trunk ENI.
+// 2) The assignment_id is not in the given running task list.
+// 3) The assignment has been created for more than 5 minutes.
+// 4) Tha assignment is not a transition assignment.
+// 5) The gc_tombstone is not already set .
+func tombstoneUnusedAssignments(ctx context.Context, tx *sql.Tx, trunkENI string, runningTasks []string) (int64, error) {
+	rowsAffected, err := tx.ExecContext(ctx, `
+WITH unused_assignments AS
+  (SELECT assignments.id
+   FROM branch_eni_attachments
+   JOIN assignments ON branch_eni_attachments.association_id = assignments.branch_eni_association
+   WHERE trunk_eni = $1
+     AND assignments.assignment_id NOT IN
+       (SELECT unnest($2::text[]))
+     AND assignments.created_at < now() - INTERVAL '5 minutes'
+     AND NOT assignments.is_transition_assignment
+     AND assignments.gc_tombstone IS NULL)
+UPDATE assignments
+SET gc_tombstone = now()
+FROM unused_assignments
+WHERE assignments.id = unused_assignments.id
+`, trunkENI, pq.Array(runningTasks))
+	if err != nil {
+		err = fmt.Errorf("Could not tombstone unused assignments: %w", err)
+		return 0, err
+	}
+
+	n, err := rowsAffected.RowsAffected()
+	if err != nil {
+		err = fmt.Errorf("Could not find rows affected: %w", err)
+		return 0, err
+	}
+	return n, nil
+}
+
+// Set gc_tombstone time to now() on any transition assginments that match the following condition:
+// 1) The transition assignment uses the given trunk ENI.
+// 2) is_transition_assignment is true.
+// 3) The transition assignment is not used by any of the given running task.
+// 4) The transition assignment has not been used for more than 5 minutes.
+// 5) The gc_tombstone is not already set.
+func tombstoneUnusedTransitionAssignments(ctx context.Context, tx *sql.Tx, trunkENI string, runningTasks []string) (int64, error) {
+	rowsAffected, err := tx.ExecContext(ctx, `
+		WITH 
+		used_transition_assignments AS (
+			SELECT transition_assignment AS id FROM assignments WHERE assignment_id IN (SELECT unnest($2::text[]))
+		),
+		unused_transition_assignments AS (
+			SELECT assignments.id 
+			FROM assignments
+			JOIN branch_eni_attachments ON branch_eni_attachments.association_id = assignments.branch_eni_association
+			WHERE trunk_eni = $1
+			AND assignments.is_transition_assignment
+			AND NOT EXISTS (
+				SELECT id from used_transition_assignments where used_transition_assignments.id = assignments.id
+			)
+			AND assignments.transition_last_used < now() - INTERVAL '5 minutes'
```

Comment:
Until we have a way to have a unique transition assignment per instance/trunk-eni/sg, do you think it might be better for this to be some value greater than 5 mins? Could we run into a situation where we end up having multiple transition assignments created over and over again?

Reply:
Good question. Note that there are 2 time durations to be aware of in a GC process. This applies to both regular and transition assignment GC. 
* 5 minutes (here): When an assignment is unused, mark `gc_timestone` only after it hasn't been used for 5 minutes.
* 30 minutes (see the query in `getAssignmentsToRemove` function): Once an assignment is marked for GC, return it in `GCV3` response so that VPC tool will tear down the network only after 30 minutes. 

As a result, an assignment will only be deleted 35 minutes after it's unused. I think that's enough time to wait before GC.

Reply:
Sounds good!